river/tree/mondrian/mondrian_tree_nodes.py

```diff
@@ -0,0 +1,474 @@
+import typing
+
+from abc import ABC, abstractmethod
+
+from river.utils.math import log_sum_2_exp
+
+from math import log
+
+from river.tree.base import Leaf
+from river.tree.base import Branch
+
+
+class MondrianTreeLeaf(Leaf, ABC):
+    """
+    Abstract class for all types of nodes in a Mondrian Tree
+
+    Parameters
+    ----------
+    parent
+        Parent Node
+    n_features
+        Number of features of the input data
+    time
+        Split time of the node for Mondrian process
+    """
+
+    def __init__(
+            self,
+            parent: Leaf or None,
+            n_features: int,
+            time: float
+    ):
+
+        super().__init__()
+
+        # Generic Node attributes
+        self.parent = parent
+        self.time = time
+        self.is_leaf = True
+        self.depth = 0
+        self._left = None
+        self._right = None
+        self.feature = 0
+        self.weight = 0.0
+        self.log_weight_tree = 0
+        self.threshold = 0.0
+        self.n_samples = 0
+        self.n_features = n_features
+        self.memory_range_min = [0] * n_features
+        self.memory_range_max = [0] * n_features
+
+    def copy(self, node):
+        """
+        Copies the node into the current one.
+        Parameters
+        ----------
+        node
+            Origin node to copy data from
+
+        Returns
+        -------
+
+        """
+        self.is_leaf = node.is_leaf
+        self.depth = node.depth
+        self.parent = node.parent
+        self._left = node.get_left()
+        self._right = node.get_right()
+        self.feature = node.feature
+        self.weight = node.weight
+        self.log_weight_tree = node.log_weight_tree
+        self.threshold = node.threshold
+        self.time = node.time
+
+    def set_left(self, node):
+        """
+        Set the left child node
+        Parameters
+        ----------
+        node
+            Child node to be set as left node
+
+        Returns
+        -------
+
+        """
+        self._left = node
+
+    def set_right(self, node):
+        """
+        Set the right child node
+        Parameters
+        ----------
+        node
+            Child node to be set as right node
+
+        Returns
+        -------
+
+        """
```

Comment:
Python properties are the preferred strategy to use :D

Reply:
I tried something different then using `property` and `variable.setter`. I'm not really an expert of OOP in Python, let me know if it's ok to you =)
It makes the code a lot more readable actually!